impedance-control/joint_impedance_cartesian_circle_motion.script

```diff
@@ -0,0 +1,141 @@
+
+    radius = 0.15         # Circle radius (meters)
+    rad_per_sec = 1.5     # Angular speed (rad/s)
+    speed_ramp_up = 0.05  # Speed ramp-up increment
+
+elif reach < 1.0:
+    # UR5e
+    max_torque = max_torque_ur3_spec / 2             # Max torque to clamp to
+    Kp = [1000, 1000, 500, 100, 100, 100]            # Proportional gains (tune as needed)
+    Kd = [2*sqrt(Kp[0]), 2*sqrt(Kp[1]), 2*sqrt(Kp[2]), 2*sqrt(Kp[3]), 2*sqrt(Kp[4]), 2*sqrt(Kp[5])]  # Derivative gains
+
+    radius = 0.25         # Circle radius (meters)
+    rad_per_sec = 1.5     # Angular speed (rad/s)
+    speed_ramp_up = 0.005 # Speed ramp-up increment
+
+else:
+    # UR10e or longer
+    max_torque = [56.0, 56.0, 28.0, 9.0, 9.0, 9.0]   # Max torque to clamp to
+    Kp = [1000, 1000, 500, 100, 100, 100] * 1.3      # Proportional gains (tune as needed)
+    Kd = [2*sqrt(Kp[0]), 2*sqrt(Kp[1]), 2*sqrt(Kp[2]), 2*sqrt(Kp[3]), 2*sqrt(Kp[4]), 2*sqrt(Kp[5])] * 1.3 # Derivative gains
+
+    radius = 0.35         # Circle radius (meters)
+    rad_per_sec = 1.5     # Angular speed (rad/s)
+    speed_ramp_up = 0.005 # Speed ramp-up increment
+end
+
+# --- Utility Functions ---
+
+# Clamp each element of an array to +/- clampValue
+def clampArray(value, clampValue):
+    pdiff = value - clampValue
+    mdiff = value + clampValue
+    ret = value
+    j = 0
+    while j < length(value):
+        if pdiff[j] > 0:
+            ret[j] = clampValue[j]
+        elif mdiff[j] < 0:
+            ret[j] = -clampValue[j]
+        end
+        j = j + 1
+    end
+    return ret
+end
+
+# Transform velocity vector v from offset t to base frame
+def screw_trans(t, v):
+  local vw = wrench_trans(t, [v[3], v[4], v[5], v[0], v[1], v[2]])
+  return [vw[3], vw[4], vw[5], vw[0], vw[1], vw[2]]
+end
+
+# --- Trajectory Generation ---
+
+# Generate circle trajectory pose, velocity, and acceleration at given time
+def get_circle_pos(time):
+    circ_pose = p[
+        radius * cos(rad_per_sec * time),
+        radius * sin(rad_per_sec * time),
+        0, 0, 0, 0
+    ]
+    circ_vel = [
+        radius * -sin(rad_per_sec * time),
+        radius * cos(rad_per_sec * time),
+        0, 0, 0, 0
+    ]
+    circ_acc = [
+        radius * -cos(rad_per_sec * time),
+        radius * -sin(rad_per_sec * time),
+        0, 0, 0, 0
+    ]
+    return struct(p = circ_pose, v = circ_vel, a = circ_acc)
+end
+
+# --- Initialization ---
+
+time = 0  # Trajectory time
+start_pos = get_target_tcp_pose()  # Initial TCP pose
+offset_pos = pose_trans(start_pos, pose_inv(get_circle_pos(time).p))  # Offset to keep circle relative to start
+
+speed_scale = 0                   # Velocity scaling factor (for smooth start)
+actual_speed_scale_acc = 0        # Actual speed scale acceleration
+q_err = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]      # Joint position error (init)
+q_err_acc = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]  # Joint acceleration error (init)
+
+# --- Main Control Loop ---
+
```

Comment:
Is there a specific initial joint configuration, that the robot should be in?

Reply:
No, It uses the current position as the start point.